Sources/Datadog/RUM/RUMVitals/VitalRefreshRateReader.swift

```diff
@@ -0,0 +1,112 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import UIKit
+
+/// A class reading the refresh rate (frames per second) of the main screen
+internal class VitalRefreshRateReader {
+    private var observers = [VitalObserver]()
+    private var displayLink: CADisplayLink?
+    private var lastFrameTimestamp: CFTimeInterval?
+    private(set) var isRunning = false
+
+    init(notificationCenter: NotificationCenter = .default) {
+        notificationCenter.addObserver(
+            self,
+            selector: #selector(appWillResignActive),
+            name: UIApplication.willResignActiveNotification,
+            object: nil
+        )
+        notificationCenter.addObserver(
+            self,
+            selector: #selector(appDidBecomeActive),
+            name: UIApplication.didBecomeActiveNotification,
+            object: nil
+        )
+    }
+
+    deinit {
+        stop()
+    }
+
+    /// `VitalRefreshRateReader` keeps pushing data to its `observers` at every new frame.
+    /// - Parameter observer: receiver of refresh rate per frame.
+    func register(_ observer: VitalObserver) {
+        DispatchQueue.main.async {
+            self.observers.append(observer)
+        }
+    }
+
+    /// `VitalRefreshRateReader` stops pushing data to `observer` once unregistered.
+    /// - Parameter observer: already added observer; otherwise nothing happens.
+    func unregister(_ observer: VitalObserver) {
+        DispatchQueue.main.async {
+            self.observers.removeAll { existingObserver in
+                return existingObserver === observer
+            }
+        }
+    }
+
+    /// Starts listening to frame paints.
+    /// - Throws: only if `UIScreen.main` cannot generate its `CADisplayLink`
+    func start() throws {
+        try private_start()
+        isRunning = true
```

Comment:
This `start()` method calls `private_start()`, which calls `stop()`, which sets `isRunning` to `false`, but then here we set it back to `true`. This logic is too much coupled and it has a hidden bug:

- 1/ `refreshRateReader.start()` → `isRunning = true`
- 2/ `appDidBecomeActive()` → it will call `private_start()` → `stop()` and leave `isRunning = false`
- 3/ `appWillResignActive()` → no change to `isRunning`
- 4/ `appDidBecomeActive()`- as `isRunning == false` it won't restart receiving updates.

This scenario corresponds to initializing SDK early in App Delegate. The refresh reader will be started and there will be one "did become active" notification delivered right away. I did this test in https://github.com/DataDog/dd-sdk-ios/commit/543b9a0288bd349e5d4945a8672d9d0a87c069ec

Reply:
i added `private_start/stop()` later and `start()` was supposed to call `private_stop()` but i forgot to change 👍 
now this logic is removed as `Reader` is made always running ✅